### PR TITLE
yongchao: Complete task

### PR DESCRIPTION
**Pull request by pycxu**

### CODE REVIEW

**Review by Crazyorchid — CHANGES_REQUESTED** [review]
rewrite requested

**Comment by NorrisWu0** [comment]
Okay, apart from what Max has requested. You probably will want to check out how to write a commit message.

**Review by NorrisWu0 — CHANGES_REQUESTED** [review]
Either explain what you have done or re-write this in your own code.

**Comment by pycxu** [comment]
> Either explain what you have done or re-write this in your own code.

I will rewrite the task as requested, create the backend first and try to write proper commit messages.

**Comment by pycxu** [comment]
Created a backend server from scratch and refactored the react code with components.

**Comment by Crazyorchid** [comment]
> Created a backend server from scratch and refactored the react code with components.

Please use the backend template we provided and edited and create a new branch and edit on it. Do not commit node_modules. Check out .gitignore on what you should not push to the remote.

File: src/App.tsx [hunk 1]
```diff
@@ -16,91 +16,42 @@ function App() {
 
   const API_URL = "http://localhost:5001/todos/";
```
Comment (Crazyorchid) [inline src/App.tsx:17]: RESTful please. 5001?

File: src/App.tsx [hunk 2]
```diff
@@ -1,53 +1,57 @@
...
+      task: '',
+      completed: false
+    }
+  ]);
 
-  const submitForm = (event: any) => {
-    event.preventDefault();
+  const API_URL = "http://localhost:5001/todos/";
+
+  // fetch all todos and set todos state
```
Comment (NorrisWu0) [inline src/App.tsx:19]: API functions can be decoupled into a separate file

File: src/App.tsx (hunk 1 above)
Comment (NorrisWu0) [reply src/App.tsx:17]: Try to think of the backend has multiple endpoints, don't just use one constant for each endpoint.

File: src/components/header/AppHeader.jsx [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import './appHeader.scss';
+
+const AppHeader = () => {
```
Comment (NorrisWu0) [inline src/components/header/AppHeader.jsx:3]: Having this in its own component is pretty good, nice job.

File: src/components/section/AppSection.jsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import './appSection.scss';
+
+const AppSection = (props) => {
```
Comment (NorrisWu0) [inline src/components/section/AppSection.jsx:3]: But why this is in its own component?

File: src/App.tsx [hunk 5]
```diff
@@ -1,53 +1,57 @@
...
-            <button type="submit" className="to-do-list__submit">
-              Add
-            </button>
-          </div>
-        </form>
-
-        {/* You should render your todo list down here */}
-      </section>
+      <AppHeader />
+      <AppSection>
```
Comment (NorrisWu0) [inline src/App.tsx:51]: Okay, maybe explain how this component is structured?

File: backend/server.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+app.use(cors())
+app.use(express.json({extended: false}))
+
+app.get('/todos', (req, res) => {
+    res.status(200).json(todos)
+})
+
+app.post('/todos', (req, res) => {
+    const newTodo = {
+        task: req.body.task,
```
Comment (Crazyorchid) [inline backend/server.js:18]: Please add restrictions to request body

File: backend/server.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        completed: false
+    }
+
+    todos.push(newTodo)
+    res.status(200).json(todos)
+})
+
+app.put('/todos/:id', (req, res) => {
+    const index = req.params.id
+    todos[index].completed = true
```
Comment (Crazyorchid) [inline backend/server.js:28]: This way only works for this temp frontend, what if you have a database or temp storage?

File: backend/server.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    res.status(200).json(todos)
+})
+
+app.delete('/todos/:id', (req, res) => {
+    const index = req.params.id
+    todos.splice(index, 1)
```
Comment (Crazyorchid) [inline backend/server.js:34]: Same here

**Review by Crazyorchid — CHANGES_REQUESTED** [review]
- Make sure you have proper routes and controllers if you want to setup your own server. 
- Please do not commit `node_modules`  (e.g. using `.gitignore`)
- Make sure you are using RESTful API formats with different endpoints
- Adding requirements to request bodies (Critical bugs on your frontend caused by this problem)

File: src/components/todo/Todo.jsx [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import './todo.scss';
+
+const Todo = ({todos, completeTask, removeTask}) => {
```
Comment (NorrisWu0) [inline src/components/todo/Todo.jsx:3]: This works, but if the component is nested deeper, you will need to pass the `completeTask` and `removeTask` multiple times. This is another good thing about decoupling the endpoint function, you can just import them when you need them.

**Comment by pycxu** [comment]
@NorrisWu0 @Crazyorchid Thanks for the review! I will think about how to better structure/decouple the frontend. I am new to backend and the model/mutiendpoint concept; I will have a look at the backend repo and try to create a new branch.

**Pull request closed without merging**